nixos/modules/services/networking/lokinet.nix

```diff
@@ -0,0 +1,161 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.lokinet;
+  settingsFormat = pkgs.formats.ini { listsAsDuplicateKeys = true; };
+  configFile = settingsFormat.generate "lokinet.ini" (lib.filterAttrsRecursive (n: v: v != null) cfg.settings);
+in with lib; {
+  options.services.lokinet = {
+    enable = mkEnableOption "Lokinet daemon.";
+
+    dataDir = mkOption {
+      type = types.str;
+      default = "/var/lib/lokinet";
+      description = "The directory where Lokinet stores its data files.";
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.lokinet;
+      defaultText = literalExpression "pkgs.lokinet";
+      description = "Lokinet package to use.";
+    };
+
+    useLocally = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = "Whether to use Lokinet locally.";
+    };
+
+    settings = mkOption {
+      type = with types;
+        submodule {
+          freeformType = settingsFormat.type;
+
+          options = {
+            dns = {
+              bind = mkOption {
+                type = str;
+                default = "127.3.2.1";
+                description = "Address to bind to for handling DNS requests.";
+              };
+
+              upstream = mkOption {
+                type = listOf str;
+                default = [ "9.9.9.10" ];
+                example = [ "1.1.1.1" "8.8.8.8" ];
+                description = ''
+                  Upstream resolver(s) to use as fallback for non-loki addresses.
+                  Multiple values accepted.
+                '';
+              };
+            };
+
+            network = {
+              exit = mkOption {
+                type = bool;
+                default = false;
+                example = true;
+                description = ''
+                  Whether to act as an exit node. Beware that this
+                  increases demand on the server and may pose liability concerns.
+                  Enable at your own risk.
+                '';
+              };
+
+              exit-node = mkOption {
+                type = nullOr (listOf str);
+                default = null;
+                example = ''
+                  exit-node = [ "example.loki" ];              # maps all exit traffic to example.loki
+                  exit-node = [ "example.loki:100.0.0.0/24" ]; # maps 100.0.0.0/24 to example.loki
+                '';
+                description = ''
+                  Specify a `.loki` address and an optional ip range to use as an exit broker.
+                  See <link xlink:href="http://probably.loki/wiki/index.php?title=Exit_Nodes"/> for
+                  a list of exit nodes.
+                '';
+              };
+
+              keyfile = mkOption {
```

Comment:
This service runs as a `DynamicUser` so by itself this is problematic. You should pair this with [LoadCredential](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Credentials).

Reply:
right now that file needs to be read/write so it can be written to with a key if the file does not exist. the key is just a 64 byte bencoded ed25519 secret key, encoded as `64:insert_64_bytes_binary_secret_here`. i didn't know about this featue in systemd, i think i will use this in the future and try to play nice with it.

Reply:
(it should be fine being read only as long as it has a valid private key in it)

Reply:
Eventually, what I can do is :
  - change the option's type to str (since the path will always be in dataDir anyway)
  - append dataDir to the value supplied by the user (with something like `crudini`) in `preStart`

Reply:
> (it should be fine being read only as long as it has a valid private key in it)

If it is a **private** key then `DynamicUser` shouldn't (directly) have access to it because that means world readable - hence `LoadCredential`.

> change the option's type to str (since the path will always be in dataDir anyway)

When you say `dataDir` you mean the hard coded string `/var/lib/loki`, right? Remember, with `DynamicUser` you can't arbitrarily choose any directory you want. You need to remove the `dataDir` option. Just checking :+1:

Reply:
> When you say `dataDir` you mean the hard coded string `/var/lib/loki`, right? Remember, with `DynamicUser` you can't arbitrarily choose any directory you want. You need to remove the `dataDir` option. Just checking +1
Yup!

Reply:
So if I understand correctly, if we're going `LoadCredential`-route, the user would have to manually generate the keyfile and we could use something similar to `passwordFile`.

Or if we want Lokinet to automatically generate the keyfile for us, then we can use my suggestion above.

Am I correct ? Just clarifying things.

Reply:
My *suggestion* would be to generate the secret if `keyPath` remains `null`, storing the key in the `StateDirectory` which your `DynamicUser` has access to. If the configuration specifies a `keyFile` then you could use `LoadCredential` to access the file... though the exact details of that I can't speak to because I'm not familiar with `loki`. Note that `LoadCredential` will pass you the new path of the secret via environment variable, something like `${CREDENTIALS_DIRECTORY}/keyfile`, which is determined at **runtime**. You would need to understand what `loki` is capable of to determine the best way to integrate `LoadCredential`.

Keep in mind this is just my suggestion. :smile:

Reply:
> My _suggestion_ would be to generate the secret if `keyPath` remains `null`

Oh, but a keyfile isn't *necessary* for Lokinet to work. It just basically allows Lokinet to keep the same .loki address each time Lokinet is restarted. It's useful e.g. when you're hosting a hidden service.

Reply:
Oh... Maybe you need an option outside the `settings` then. Something like "generateKeyOnFirstRun" or... I don't know :man_shrugging:

I feel like I'd need to know more about the application to add any value to this conversation at this point... so I'll humbly bow out :bowing_man:

Reply:
> If it is a **private** key then `DynamicUser` shouldn't (directly) have access to it because that means world readable - hence `LoadCredential`.

I just have a little question regarding this. In my experience, I can't see the content of `/var/lib/lokinet` (with DynamicUser) at all except if I'm the root user. How is that "world readable" ?
If I can't see at all any content in `/var/lib/lokinet` without being root, I don't understand the point in using `LoadCredential`. Maybe I'm misunderstanding something.

Reply:
You understand correctly it is just that generally it is more flexible to put secrets elsewhere. The idea is that maybe you have a mechanism of retrieving secrets that is unable to write to that directory, etc...

At this point I think you should abandon my comments and do what you were doing, though :laughing: Sorry for the hassle, I didn't realize it would require more insight/work to provide a complete solution.

Reply:
No problem! I appreciate you giving your time. :heart: